compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1750,7 +1750,24 @@ impl<'a> Parser<'a> {
     /// Parses an `if` expression (`if` token already eaten).
     fn parse_if_expr(&mut self, attrs: AttrVec) -> PResult<'a, P<Expr>> {
         let lo = self.prev_token.span;
-        let cond = self.parse_cond_expr()?;
+        let cond = self.parse_cond_expr()?.into_inner();
+
+        let cond = if let ExprKind::Paren(paren) = &cond.kind {
```

Comment:
This should only be done if the gate is disabled, otherwise there's no way to accept this syntax ever, which is valid when let chains are enabled.

Reply:
Hm.. in the current version, we lower `if (let ...)` to `if let` unconditionally, so that's what I did here as well.

> This should only be done if the gate is disabled

Which gate do you mean?

Reply:
`let_chains`. I think you can access it with `self.sess.features_untracked().or_patterns`. If that expr is true, then we shouldn't emit this error. I think this is a bug introduced by me when I added the targeted error.

Reply:
If we do this, then parsing of `if (let ...)` will be different depending on whether `let_chains` is enabled or not, right? Should feature gates be able to change parsing? I'd expect not.

Reply:
The feature gates would change what the resulting AST is interpreted as. `if (let pat = expr)` _is valid_ when the feature gate is enabled. This is why you get the error we're trying to silence. All of this is _appropriate behavior_, but it makes for a bad _experience_. Because of this ideally the feature gate error would suggest removing the parens for this case, but that would mean ferrying more information into the linting machinery so that it can know that it was this case. I find it easier to treat it more locally.

Reply:
I think it's not possible to check for gates in parser. The data is not available in `ParseSess`.

I think for this information to be available when parsing you would need to first parse the crate root for the attributes, then parse other files. Perhaps the parser is not implemented this way currently? (though I just saw the `Edition` field, which can be updated with an attribute at the crate root..)